### PR TITLE
MelBandRoformer: kernelized FP8 audio source separation pipeline

### PR DESCRIPTION
**Pull request by xym808**
## MelBandRoformer — kernelized FP8 inference for audio source separation

Full kernelized FP8 inference for **MelBandRoformer**, an audio source
separation model (vocals / background) with a time-frequency dual-band
Transformer backbone (6 time + 6 freq layers, dim=384, 60 mel bands, 8 heads).

The default hot path runs **FP8 E4M3 GEMM** (cuBLASLt), **Flash SDPA**
attention, and **eight custom fused CUDA kernels** — no native PyTorch matmul
or elementwise norm in the Transformer compute path.

## Adds

- `flash_rt/models/melband_roformer/` — `MelBandRoformerPipeline` with FP8
  weight pre-quantization, offline activation calibration, and weight-deletion
  memory optimization (~400 MB freed after quantization).
- `csrc/kernels/mbr_kernels.{cu,cuh}` — eight fused BF16/FP8 kernels.
- `tests/test_melband_roformer_smoke.py` — import / API / gated-symbol smoke
  test (no checkpoint required).
- `docs/melband_roformer_usage.md` — build, kernel reference, performance.

## Custom fused kernels

| Kernel | Fused operation |
|--------|----------------|
| `mbr_qkv_split_rope` | QKV split + interleaved RoPE → (B,H,S,D) |
| `mbr_gated_attn_quant` | sigmoid gate · attn + reshape + FP8 quant |
| `mbr_fp8_dequant_bf16` | FP8 E4M3 → BF16 dequantize |
| `mbr_resadd_rmsnorm_fp8_keepres` | residual add + RMSNorm → FP8 (keeps BF16 residual) |
| `mbr_fused_add_rmsnorm_bf16` | residual add + RMSNorm (BF16 in/out) |
| `mbr_fp8_dequant_tiny_linear_bf16` | FP8 dequant + tiny linear (for gates) |
| `mbr_fp8_gemm_bias_residual_norm_bf16` | FP8 GEMM + bias + residual + RMSNorm |
| `mbr_gemm_output_bias_residual_norm_bf16` | GEMM output + bias + residual + RMSNorm |

All kernels follow FlashRT conventions: `template<typename T>`, `common.cuh`
reuse (`to_f32`, `from_f32`, `packed2`, `block_reduce_sum`), typed launchers,
`typed_ptr<T>()` bindings.

## Compute path (all FlashRT / cuBLASLt kernels)

| Stage | Kernel |
|-------|--------|
| Norm → FP8 | `fvk.rms_norm_fp8` |
| FP8 GEMM (qkv, out, ffn1, ffn2) | `torch._scaled_mm` (cuBLASLt) |
| QKV split + RoPE | `mbr_qkv_split_rope` |
| Attention | Flash SDPA (`F.scaled_dot_product_attention`) |
| Gate + quant | `mbr_gated_attn_quant` |
| Residual + norm → FP8 | `mbr_resadd_rmsnorm_fp8_keepres` |
| GeLU + bias + quant | `fvk.bias_gelu_quantize_fp8_static_bf16` |
| Final residual + norm | `mbr_fused_add_rmsnorm_bf16` |

## Performance (RTX 5060 Ti, SM120, CUDA 13)

120 s stereo audio (44.1 kHz), overlap-add (num_overlap=2), batch_size=4:

| Metric | Baseline (PyTorch) | Optimized (FP8) |
|--------|--------------------|-----------------|
| **RTF** | **0.077** | **0.025** |
| **Speedup** | 1.0× | **3.1×** |
| Peak VRAM | 1056 MB | 1507 MB |
| **Cosine similarity** | — | **0.9997** |
| **Max abs diff** | — | **0.074** |

RTF = inference time / audio duration. Baseline is vanilla
`melband-roformer-infer` (no FlashRT). Cosine and max abs are computed
sample-by-sample on the full vocals output vs the baseline.

## Precision

FP8 E4M3 activation scales are determined by offline calibration
(`fp8_calibration.json`, 176 per-layer static scales). FP8 weights are
pre-quantized at load time (`weight / scale → clamp → fp8`). The pipeline
deletes original BF16 Linear weights after quantization to save ~400 MB VRAM.

## Build isolation

All MelBandRoformer kernels are gated behind `FLASHRT_ENABLE_MELBAND_ROFORMER`
(default OFF) **and** `FLASHRT_HAVE_MELBAND_ROFORMER`, so SM89 / SM87 / SM110
and non-MelBand builds never compile them; bindings are
`#ifdef FLASHRT_HAVE_MELBAND_ROFORMER`. Build with
`-DFLASHRT_ENABLE_MELBAND_ROFORMER=ON`. **csrc is additions-only vs main
(0 deletions).**

## Integration

- `flash_rt/models/melband_roformer/` follows the existing model directory
  convention (`__init__.py` + `pipeline.py`), analogous to `nexn2/`, `pi05/`.
- The pipeline loads the model via `melband-roformer-infer` (unmodified PyPI
  package) and replaces the Transformer forward at runtime via monkeypatch —
  no package source edits.
- Model weights: `poiqazwsx/melband-roformer-denoise` (HuggingFace) or
  `cpadyun/melband-roformer` (ModelScope).

## Test

```bash
# Build
cmake .. -DFLASHRT_ENABLE_MELBAND_ROFORMER=ON && make -j$(nproc)

# Smoke test (no GPU / checkpoint needed)
pytest tests/test_melband_roformer_smoke.py
```

## Hardware

RTX 5060 Ti / SM120, CUDA 13, Python 3.12.

### CODE REVIEW

**Comment by LiangSu8899** [comment]
Thanks for the PR. The CMake-level isolation looks good: the MelBandRoformer kernels are gated behind `FLASHRT_ENABLE_MELBAND_ROFORMER=ON`, default builds do not compile them, and the gated SM120 build links successfully.

Before merge, please address the following items.

## Required fixes

1. **Make Python import safe when MBR kernels are not built**

   `flash_rt.models.melband_roformer.pipeline` currently imports `flash_rt_kernels` and binds `fvk.mbr_*` at module import time.

   Since `FLASHRT_ENABLE_MELBAND_ROFORMER` is OFF by default, this makes:

   ```python
   from flash_rt.models.melband_roformer import MelBandRoformerPipeline
   ```

   fail in a normal checkout/default build.

   Please change this to lazy validation:

   - package/module import should not require `flash_rt_kernels`
   - `MelBandRoformerPipeline.__init__` should require the kernels
   - if symbols are missing, raise a clear `RuntimeError` mentioning:
     `-DFLASHRT_ENABLE_MELBAND_ROFORMER=ON`

2. **Remove temporary device scale tensors from the hot path**

   The forward currently creates temporary tensors and immediately passes their `data_ptr()` into async CUDA kernels, for example:

   ```python
   torch.tensor([w2.out_scale], dtype=torch.float32, device=x.device).data_ptr()
   ```

   This is unsafe because the tensor may be freed after the expression while the kernel is still using the pointer.

   Please allocate persistent device scale tensors during weight setup, e.g. in `_W`:

   ```python
   self.out_scale_dev = torch.tensor([out_scale], dtype=torch.float32, device=device)
   ```

   Then use `w.out_scale_dev.data_ptr()` in the hot path.

3. **Fix the smoke tests**

   The smoke tests currently fail in a default build. Please ensure tests cover:

   - default build / no MBR kernels: package import succeeds
   - constructing the pipeline without MBR kernels fails fast with a clear RuntimeError
   - gated build: all required `mbr_*` symbols are present

4. **Clean kernel comments**

   Please remove development-history comments like `V70`, `V72`, `V73 ULTIMATE`, and informal/non-English notes from `mbr_kernels.cu/.cuh`.

   Keep comments factual and stable: what the kernel does, input/output layout, dtype assumptions, and constraints.

5. **Align docs, tests, and exported symbols**

   The bindings export more MBR symbols than the docs/tests list. Please either:

   - document and test every exported `mbr_*` symbol, or
   - remove unused/experimental exported kernels from this PR

## Checklist before re-review

- [ ] Branch rebased/merged onto latest `origin/main`
- [ ] `git diff --check origin/main...HEAD` passes
- [ ] `python -m compileall flash_rt/models/melband_roformer tests/test_melband_roformer_smoke.py` passes
- [ ] `pytest -q tests/test_melband_roformer_smoke.py` passes in a default build
- [ ] `GPU_ARCH=120 FLASHRT_ENABLE_MELBAND_ROFORMER=ON` build of `flash_rt_kernels` passes
- [ ] Gated build exports all required MBR symbols
- [ ] No private paths, local environment notes, debug prints, or temporary benchmark scratch text
- [ ] No temporary CUDA tensor `data_ptr()` passed to async kernels
- [ ] Docs accurately describe build flag, runtime requirements, supported GPU arch, and expected usage

**Review by LiangSu8899 — APPROVED** [review]
